File: typescript/src/server/files.ts
import type { YirTransport, YirRequestOptions } from "./client.js";
import { YirSDKValidationError } from "../shared/standard.js";

export type CreateFile = { readonly name: string; readonly media_type: string; readonly size: number };
export type CreateFilesRequest = { readonly files: readonly CreateFile[]; readonly upload_mode?: "auto" | "multipart" };
export type InputFile = CreateFile & {
  readonly id: string;
  readonly object: "file";
  readonly status: "pending_upload" | "ready" | "expired" | "failed";
  readonly url?: string;
  readonly expires_at?: number;
  readonly upload?: {
    readonly type: "multipart";
    readonly expires_at: number;
    readonly parts: readonly { readonly part_number: number; readonly size: number; readonly url: string }[];
  };
};

export type YirFileClient = {
  createFiles(request: CreateFilesRequest, idempotencyKey: string, options?: YirRequestOptions): Promise<readonly InputFile[]>;
  getFile(id: string, options?: YirRequestOptions): Promise<InputFile>;
  completeFile(id: string, options?: YirRequestOptions): Promise<InputFile>;
};

const fileIDPattern = /^file_[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const allowedMediaTypes = new Set(["image/jpeg", "image/png", "image/webp", "video/mp4", "video/webm", "audio/mpeg", "audio/wav"]);

export function createFileClient(transport: YirTransport): YirFileClient {
  return {
    async createFiles(request, idempotencyKey, options) {
      options?.signal?.throwIfAborted();
      const key = idempotencyKey.trim();
      if (!key || new TextEncoder().encode(key).length > 255 || /[\r\n]/.test(key)) invalid("idempotency_key_invalid", "idempotencyKey");
      if (!request || !Array.isArray(request.files) || request.files.length < 1 || request.files.length > 20) invalid("file_count_invalid", "files");
      if (request.upload_mode !== undefined && request.upload_mode !== "auto" && request.upload_mode !== "multipart") invalid("upload_mode_invalid", "upload_mode");
      for (const field of Object.keys(request)) if (!["files", "upload_mode"].includes(field)) invalid("unknown_field", field);
      request = { ...request, files: request.files.map((file, index) => normalizeFileMetadata(file, `files.${index}`)) };
      const result = await transport<{ files: InputFile[] }>({ method: "POST", path: "/v1/files", headers: { "Idempotency-Key": key }, body: request,
        ...(options?.signal ? { signal: options.signal } : {}) });
      if (!result || !Array.isArray(result.files) || result.files.length !== request.files.length) throw new Error("file_response_invalid");
      for (const [index, file] of result.files.entries()) {
        validateFileResponse(file);
        const expected = request.files[index]!;
        if (file.name !== expected.name || file.size !== expected.size || file.media_type !== expected.media_type) throw new Error("file_response_invalid");
      }
      return result.files;
    },
    getFile(id, options) { return fileRequest(transport, id, false, options); },
    completeFile(id, options) { return fileRequest(transport, id, true, options); },
  };
}

function validateFileMetadata(file: CreateFile, path: string) {
  if (!file || typeof file !== "object" || Array.isArray(file)) invalid("file_metadata_invalid", path);
  for (const field of Object.keys(file)) if (!["name", "media_type", "size"].includes(field)) invalid("unknown_field", `${path}.${field}`);
  if (typeof file.name !== "string" || !file.name.trim() || new TextEncoder().encode(file.name.trim()).length > 255 || /[\x00-\x1f\x7f/\\]/.test(file.name)) invalid("file_name_invalid", `${path}.name`);
  if (!allowedMediaTypes.has(file.media_type)) invalid("file_media_type_invalid", `${path}.media_type`);
  if (!Number.isSafeInteger(file.size) || file.size < 1 || file.size > 2147483648) invalid("file_size_invalid", `${path}.size`);
}

function normalizeFileMetadata(file: CreateFile, path: string): CreateFile {
  if (!file || typeof file !== "object") invalid("file_metadata_invalid", path);
  const normalized = { ...file, name: typeof file.name === "string" ? file.name.trim() : file.name,
    media_type: typeof file.media_type === "string" ? file.media_type.trim().toLowerCase() : file.media_type };
  validateFileMetadata(normalized, path);
  return normalized;
}

function validateFileResponse(file: InputFile) {
  if (!file || !fileIDPattern.test(file.id) || file.object !== "file" || !["pending_upload", "ready", "expired", "failed"].includes(file.status)) throw new Error("file_response_invalid");
}

async function fileRequest(transport: YirTransport, id: string, complete: boolean, options?: YirRequestOptions): Promise<InputFile> {
  options?.signal?.throwIfAborted();
  if (!fileIDPattern.test(id)) invalid("file_id_invalid", "id");
  const result = await transport<InputFile>({ method: complete ? "POST" : "GET", path: `/v1/files/${id}${complete ? "/complete" : ""}`,
    ...(complete ? { body: {} } : {}), ...(options?.signal ? { signal: options.signal } : {}) });
  validateFileResponse(result);
  if (result.id !== id) throw new Error("file_response_invalid");
  return result;
}

/** Reuses a ready file or streams its issued plan, then confirms completion. */
export async function uploadFile(client: Pick<YirFileClient, "completeFile">, file: InputFile, data: Blob,
  options: YirRequestOptions & { fetch?: typeof globalThis.fetch } = {}): Promise<InputFile> {
  options.signal?.throwIfAborted();
  validateFileResponse(file);
  if (!(data instanceof Blob) || data.size !== file.size || file.size < 1 || file.size > 2147483648) invalid("upload_source_invalid", "data");
  if (file.status === "ready") return file;
  const plan = file.upload;
  if (!(data instanceof Blob) || data.size !== file.size || file.size < 1 || file.size > 2147483648 || file.status !== "pending_upload" ||
    !plan || plan.type !== "multipart" || !Number.isSafeInteger(plan.expires_at) || plan.expires_at <= Math.floor(Date.now() / 1000) || !Array.isArray(plan.parts) || !plan.parts.length) invalid("upload_plan_invalid", "upload");
  let total = 0;
  for (const [index, part] of plan.parts.entries()) {
    if (part.part_number !== index + 1 || !Number.isSafeInteger(part.size) || part.size <= 0 || part.size > file.size - total) invalid("upload_plan_invalid", "upload.parts");
    try {
      const url = new URL(part.url);
      if (url.protocol !== "https:" || url.username || url.password || url.hash) throw new Error();
    } catch { invalid("upload_plan_invalid", "upload.parts"); }
    total += part.size;
  }
  if (total !== file.size) invalid("upload_plan_invalid", "upload.parts");
  const fetchPart = options.fetch ?? globalThis.fetch;
  let offset = 0;
  for (const part of plan.parts) {
    let response: Response;
    try {
      response = await fetchPart(part.url, { method: "PUT", body: data.slice(offset, offset + part.size), signal: options.signal,
        redirect: "error", credentials: "omit" });
    } catch {
      options.signal?.throwIfAborted();
      // Network errors can contain the signed upload URL; never echo them.
      throw new Error("upload_transport_failed");
    }
    await response.body?.cancel();
    if (!response.ok) throw new Error(`upload_failed:${response.status}`);
    offset += part.size;
  }
  const ready = await client.completeFile(file.id, options);
  validateFileResponse(ready);
  if (ready.status !== "ready" || ready.id !== file.id || ready.name !== file.name || ready.media_type !== file.media_type || ready.size !== file.size) throw new Error("file_response_invalid");
  return ready;
}

/** Persist key before calling; retries must use identical metadata and bytes. */
export async function createAndUploadFile(client: Pick<YirFileClient, "createFiles" | "completeFile">,
  metadata: CreateFile, data: Blob, key: string,
  options: YirRequestOptions & { fetch?: typeof globalThis.fetch } = {}): Promise<InputFile> {
  options.signal?.throwIfAborted();
  metadata = normalizeFileMetadata(metadata, "file");
  if (!(data instanceof Blob) || data.size !== metadata.size) invalid("upload_source_invalid", "data");
  const files = await client.createFiles({ files: [metadata], upload_mode: "multipart" }, key, options);
  if (files.length !== 1) throw new Error("file_response_invalid");
  const file = files[0]!;
  validateFileResponse(file);
  if (file.name !== metadata.name || file.media_type !== metadata.media_type || file.size !== metadata.size) throw new Error("file_response_invalid");
  return uploadFile(client, file, data, options);
}

function invalid(code: string, path: string): never { throw new YirSDKValidationError(code, path); }

File: go/files.go
package yir

import (
	"context"
	"errors"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
	"unicode/utf8"
)

type CreateFile struct {
	Name      string `json:"name"`
	MediaType string `json:"media_type"`
	Size      int64  `json:"size"`
}

type CreateFilesRequest struct {
	Files      []CreateFile `json:"files"`
	UploadMode string       `json:"upload_mode,omitempty"`
}

type File struct {
	ID        string      `json:"id"`
	Object    string      `json:"object"`
	Status    string      `json:"status"`
	Name      string      `json:"name"`
	MediaType string      `json:"media_type"`
	Size      int64       `json:"size"`
	Upload    *FileUpload `json:"upload,omitempty"`
	URL       string      `json:"url,omitempty"`
	ExpiresAt int64       `json:"expires_at,omitempty"`
}

type FileUpload struct {
	Type      string       `json:"type"`
	Parts     []UploadPart `json:"parts"`
	ExpiresAt int64        `json:"expires_at"`
}

type UploadPart struct {
	PartNumber int    `json:"part_number"`
	Size       int64  `json:"size"`
	URL        string `json:"url"`
}

// CreateAndUploadFile prepares one replayable input file. Persist key before
// calling and reuse it with the same metadata and bytes after an interrupted call.
// A ready replay is returned without uploading again; failed/expired files require
// a new preparation rather than silently changing this idempotent identity.
func (c *Client) CreateAndUploadFile(ctx context.Context, metadata CreateFile, source io.ReaderAt, key string) (File, error) {
	if source == nil {
		return File{}, errors.New("upload_source_required")
	}
	metadata = normalizeCreateFile(metadata)
	files, err := c.CreateFiles(ctx, CreateFilesRequest{Files: []CreateFile{metadata}, UploadMode: "multipart"}, key)
	if err != nil {
		return File{}, err
	}
	file := files[0]
	switch file.Status {
	case "ready":
		return file, nil
	case "pending_upload":
		ready, err := c.UploadFile(ctx, file, source)
		if err != nil {
			return File{}, err
		}
		if ready.Status != "ready" || ready.Name != metadata.Name || ready.MediaType != metadata.MediaType || ready.Size != metadata.Size {
			return File{}, errors.New("response_invalid")
		}
		return ready, nil
	default:
		return File{}, errors.New("file_not_ready")
	}
}

func (c *Client) CreateFiles(ctx context.Context, request CreateFilesRequest, key string) ([]File, error) {
	key = strings.TrimSpace(key)
	if key == "" || len(key) > 255 || strings.ContainsAny(key, "\r\n") {
		return nil, errors.New("idempotency_key_invalid")
	}
	if len(request.Files) == 0 || len(request.Files) > 20 {
		return nil, errors.New("file_count_invalid")
	}
	if request.UploadMode != "" && request.UploadMode != "auto" && request.UploadMode != "multipart" {
		return nil, errors.New("upload_mode_invalid")
	}
	request.Files = append([]CreateFile(nil), request.Files...)
	for index := range request.Files {
		file := normalizeCreateFile(request.Files[index])
		request.Files[index] = file
		if file.Name == "" || !utf8.ValidString(file.Name) || len(file.Name) > 255 || strings.ContainsAny(file.Name, `/\`) ||
			strings.IndexFunc(file.Name, func(r rune) bool { return r < 0x20 || r == 0x7f }) >= 0 || file.Size < 1 || file.Size > 2147483648 {
			return nil, errors.New("file_metadata_invalid")
		}
		switch file.MediaType {
		case "image/jpeg", "image/png", "image/webp", "video/mp4", "video/webm", "audio/mpeg", "audio/wav":
		default:
			return nil, errors.New("file_media_type_invalid")
		}
	}
	var result struct {
		Files []File `json:"files"`
	}
	if err := c.do(ctx, http.MethodPost, "/v1/files", key, request, &result); err != nil {
		return nil, err
	}
	if len(result.Files) != len(request.Files) {
		return nil, errors.New("response_invalid")
	}
	for i, file := range result.Files {
		if !validFile(file) || file.Name != request.Files[i].Name || file.MediaType != request.Files[i].MediaType || file.Size != request.Files[i].Size {
			return nil, errors.New("response_invalid")
		}
	}
	return result.Files, nil
}

func normalizeCreateFile(file CreateFile) CreateFile {
	file.Name = strings.TrimSpace(file.Name)
	file.MediaType = strings.ToLower(strings.TrimSpace(file.MediaType))
	return file
}

func validFile(file File) bool {
	return fileIDPattern.MatchString(file.ID) && file.Object == "file" &&
		(file.Status == "pending_upload" || file.Status == "ready" || file.Status == "expired" || file.Status == "failed")
}

func (c *Client) GetFile(ctx context.Context, id string) (File, error) {
	return c.fileRequest(ctx, http.MethodGet, id, "", nil)
}

func (c *Client) CompleteFile(ctx context.Context, id string) (File, error) {
	return c.fileRequest(ctx, http.MethodPost, id, "/complete", struct{}{})
}

func (c *Client) fileRequest(ctx context.Context, method, id, suffix string, body any) (File, error) {
	var file File
	if !fileIDPattern.MatchString(id) {
		return file, errors.New("file_id_invalid")
	}
	err := c.do(ctx, method, "/v1/files/"+id+suffix, "", body, &file)
	if err == nil && (!validFile(file) || file.ID != id) {
		err = errors.New("response_invalid")
	}
	return file, err
}

// UploadFile streams a previously created upload plan, then asks Gateway to
// verify completion. It never forwards the Gateway Bearer key to upload URLs.
// ReaderAt supports large local files without buffering them in memory.
func (c *Client) UploadFile(ctx context.Context, file File, source io.ReaderAt) (File, error) {
	if c == nil || c.httpClient == nil {
		return file, errors.New("client_uninitialized")
	}
	if source == nil || !validFile(file) || file.Status != "pending_upload" || file.Upload == nil || file.Upload.Type != "multipart" ||
		file.Size < 1 || file.Size > 2147483648 || len(file.Upload.Parts) == 0 || file.Upload.ExpiresAt <= time.Now().Unix() {
		return file, errors.New("upload_plan_invalid")
	}
	var total int64
	for i, part := range file.Upload.Parts {
		u, err := url.Parse(part.URL)
		if err != nil || u.Scheme != "https" || u.Hostname() == "" || u.User != nil || u.Fragment != "" || part.PartNumber != i+1 || part.Size < 1 || part.Size > file.Size-total {
			return file, errors.New("upload_plan_invalid")
		}
		total += part.Size
	}
	if total != file.Size {
		return file, errors.New("upload_plan_invalid")
	}
	var offset int64
	for _, part := range file.Upload.Parts {
		request, err := http.NewRequestWithContext(ctx, http.MethodPut, part.URL, io.NewSectionReader(source, offset, part.Size))
		if err != nil {
			return file, errors.New("upload_plan_invalid")
		}
		request.ContentLength = part.Size
		result, err := c.httpClient.Do(request)
		if err != nil {
			if ctx.Err() != nil {
				return file, ctx.Err()
			}
			// Transport errors may contain the signed URL. Do not expose it.
			return file, errors.New("upload_transport_failed")
		}
		io.Copy(io.Discard, io.LimitReader(result.Body, 65536))
		result.Body.Close()
		if result.StatusCode < 200 || result.StatusCode >= 300 {
			return file, &APIError{Status: result.StatusCode, Code: "upload_failed"}
		}
		offset += part.Size
	}
	ready, err := c.CompleteFile(ctx, file.ID)
	if err != nil {
		return File{}, err
	}
	if ready.Status != "ready" || ready.Name != file.Name || ready.MediaType != file.MediaType || ready.Size != file.Size {
		return File{}, errors.New("response_invalid")
	}
	return ready, nil
}
